runtime/lua/vim/lsp/_folding_range.lua

```diff
@@ -0,0 +1,404 @@
+    return
+  end
+
+  -- Register the new `bufstate`.
+  bufstates[bufnr] = {
+    client_ranges = {},
+    row_level = {},
+    row_kinds = {},
+    row_text = {},
+  }
+
+  -- Event hooks from `buf_attach` can't be removed externally.
+  -- Hooks and `bufstate` share the same lifecycle;
+  -- they should self-destroy if `bufstate == nil`.
+  api.nvim_buf_attach(bufnr, false, {
+    -- `on_detach` also runs on buffer reload (`:e`).
+    -- Ensure `bufstate` and hooks are cleared to avoid duplication or leftover states.
+    on_detach = function()
+      bufstates[bufnr] = nil
+      api.nvim_clear_autocmds({ buffer = bufnr, group = augroup_setup })
+    end,
+    -- Reset `bufstate` and request folding ranges.
+    on_reload = function()
+      bufstates[bufnr] = {
+        client_ranges = {},
+        row_level = {},
+        row_kinds = {},
+        row_text = {},
+      }
+      for _, client in
+        ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+      do
+        request(bufnr, client)
+      end
+    end,
+    --- Sync changed rows with their previous foldlevels before applying new ones.
+    on_bytes = function(_, _, _, start_row, _, _, old_row, _, _, new_row, _, _)
+      if bufstates[bufnr] == nil then
+        return true
+      end
+      local row_level = bufstates[bufnr].row_level
+      if next(row_level) == nil then
+        return
+      end
+      local row = new_row - old_row
+      if row > 0 then
+        vim._list_insert(row_level, start_row, start_row + math.abs(row) - 1, { -1 })
+        -- If the previous row ends a fold,
+        -- Nvim treats the first row after consecutive `-1`s as a new fold start,
+        -- which is not the desired behavior.
+        local prev_level = row_level[start_row - 1]
+        if prev_level and prev_level[2] == '<' then
+          row_level[start_row] = { prev_level[1] - 1 }
+        end
+      elseif row < 0 then
+        vim._list_remove(row_level, start_row, start_row + math.abs(row) - 1)
+      end
+    end,
+  })
+  api.nvim_create_autocmd('LspDetach', {
+    group = augroup_setup,
+    buffer = bufnr,
+    callback = function(args)
+      ---@type integer
+      local client_id = args.data.client_id
+      bufstates[bufnr].client_ranges[client_id] = nil
+
+      ---@type vim.lsp.Client[]
+      local clients = vim
+        .iter(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+        ---@param client vim.lsp.Client
+        :filter(function(client)
+          return client.id ~= client_id
+        end)
+        :totable()
+      if #clients == 0 then
+        if api.nvim_buf_is_loaded(bufnr) then
+          bufstates[bufnr] = {
+            client_ranges = {},
+            row_level = {},
+            row_kinds = {},
+            row_text = {},
+          }
+          foldupdate(bufnr)
+        end
+      else
+        for _, client in ipairs(clients) do
+          request(bufnr, client)
+        end
```

Comment:
I don't quite understand why it's necessary to re-request the fold information for other clients if one client detaches from the buffer. The state should still be up to date, no?

Shouldn't this instead only clear out the `bufstate.client_ranges` for the given client id?

Reply:
Yes, this can be done; I didn't add `client_ranges` to handle multiple clients before.